File: android/src/main/kotlin/com/atmosphere/foursquare_movement_sdk/FoursquareMovementSdkPlugin.kt
// Copyright 2026 Atmosphere Innovations, Inc. All rights reserved.
// Use of this source code is governed by an MIT-style license that can be
// found in the LICENSE file.

package com.atmosphere.foursquare_movement_sdk

import android.app.Activity
import android.content.Context
import android.content.Intent
import com.foursquare.api.FoursquareLocation
import com.foursquare.api.types.Category
import com.foursquare.api.types.Photo
import com.foursquare.api.types.Venue
import com.foursquare.api.types.geofence.GeofenceEvent
import com.foursquare.api.types.geofence.GeofenceEventType
import com.foursquare.movement.BackfillNotification
import com.foursquare.movement.Confidence
import com.foursquare.movement.CurrentLocation
import com.foursquare.movement.GeofenceEventNotification
import com.foursquare.movement.LocationType
import com.foursquare.movement.MovementSdk
import com.foursquare.movement.NotificationHandler
import com.foursquare.movement.NotificationTester
import com.foursquare.movement.UserInfo
import com.foursquare.movement.UserStateNotification
import com.foursquare.movement.Visit
import com.foursquare.movement.VisitNotification
import com.foursquare.movement.debugging.DebugActivity
import com.foursquare.api.types.Journey
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import android.os.Handler
import android.os.Looper
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/** FoursquareMovementSdkPlugin */
class FoursquareMovementSdkPlugin :
    FlutterPlugin,
    ActivityAware,
    MovementSdkHostApi,
    NotificationHandler {

    private lateinit var context: Context
    private var activity: Activity? = null
    private var flutterApi: MovementSdkFlutterApi? = null
    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
    private val mainHandler: Handler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        MovementSdkHostApi.setUp(flutterPluginBinding.binaryMessenger, this)
        flutterApi = MovementSdkFlutterApi(flutterPluginBinding.binaryMessenger)
        try {
            MovementSdk.get().setNotificationHandler(this)
        } catch (_: Exception) {
            // SDK may not be initialized yet
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        MovementSdkHostApi.setUp(binding.binaryMessenger, null)
        flutterApi = null
        executor.shutdown()
    }

    // MARK: - NotificationHandler

    override fun handleVisit(context: Context, notification: VisitNotification) {
        val converted = convertVisit(notification.visit)
        mainHandler.post { flutterApi?.onVisit(converted) {} }
    }

    override fun handleBackfillVisit(context: Context, notification: BackfillNotification) {
        val converted = convertVisit(notification.visit)
        mainHandler.post { flutterApi?.onBackfillVisit(converted) {} }
    }

    override fun handleGeofenceEventNotification(context: Context, notification: GeofenceEventNotification) {
        val converted = notification.geofenceEvents.map { convertGeofenceEvent(it) }
        mainHandler.post { flutterApi?.onGeofenceEvents(converted) {} }
    }

    override fun handleUserStateChange(context: Context, notification: UserStateNotification) {
        // Not yet exposed to Flutter
    }

    override fun handleJourneyUpdate(context: Context, journey: Journey) {
        // Not yet exposed to Flutter
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // MARK: - MovementSdkHostApi Implementation

    override fun getInstallId(callback: (Result<String>) -> Unit) {
        callback(Result.success(MovementSdk.getInstallId()))
    }

    override fun start() {
        MovementSdk.start(context)
    }

    override fun stop() {
        MovementSdk.stop(context)
    }

    override fun getCurrentLocation(callback: (Result<FSQCurrentLocation>) -> Unit) {
        executor.execute {
            val result = MovementSdk.get().getCurrentLocation()
            if (result.isOk) {
                callback(Result.success(convertCurrentLocation(result.result)))
            } else {
                callback(Result.failure(Exception(result.err)))
            }
        }
    }

    override fun isEnabled(callback: (Result<Boolean>) -> Unit) {
        callback(Result.success(MovementSdk.isEnabled()))
    }

    override fun fireTestVisit(latitude: Double, longitude: Double) {
        NotificationTester.sendTestVisitArrivalAtLocation(
            context,
            latitude,
            longitude,
            false
        )
    }

    override fun showDebugScreen() {
        val intent = Intent(context, DebugActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        context.startActivity(intent)
    }

    override fun getUserInfo(callback: (Result<Map<String, String>>) -> Unit) {
        val userInfo = MovementSdk.get().userInfo
        val result = mutableMapOf<String, String>()
        userInfo?.entries?.forEach { entry ->
            result[entry.key] = entry.value
        }
        callback(Result.success(result))
    }

    override fun setUserInfo(userInfo: Map<String, String>, persisted: Boolean) {
        val fsqUserInfo = UserInfo()
        userInfo.entries.forEach { entry ->
            fsqUserInfo[entry.key] = entry.value
        }
        MovementSdk.get().setUserInfo(fsqUserInfo, persisted)
    }

    override fun setUserId(userId: String, persisted: Boolean) {
        val fsqUserInfo = MovementSdk.get().userInfo ?: UserInfo()
        fsqUserInfo.setUserId(userId)
        MovementSdk.get().setUserInfo(fsqUserInfo, persisted)
    }

    override fun clearAllData() {
        MovementSdk.clearAllData(context)
    }

    override fun getActiveVisit(callback: (Result<FSQVisit?>) -> Unit) {
        executor.execute {
            val visit = MovementSdk.get().getCurrentVisit(context)
            callback(Result.success(visit?.let { convertVisit(it) }))
        }
    }

    // MARK: - Conversion Methods

    private fun convertCurrentLocation(currentLocation: CurrentLocation): FSQCurrentLocation {
        return FSQCurrentLocation(
            currentPlace = convertVisit(currentLocation.currentPlace),
            matchedGeofences = currentLocation.matchedGeofences.map { convertGeofenceEvent(it) }
        )
    }

    private fun convertVisit(visit: Visit): FSQVisit {
        val locationType = when (visit.getType()) {
            LocationType.HOME -> FSQLocationType.HOME
            LocationType.WORK -> FSQLocationType.WORK
            LocationType.VENUE -> FSQLocationType.VENUE
            else -> FSQLocationType.UNKNOWN
        }
        val confidence = when (visit.getConfidence()) {
            Confidence.LOW -> FSQConfidenceLevel.LOW
            Confidence.MEDIUM -> FSQConfidenceLevel.MEDIUM
            Confidence.HIGH -> FSQConfidenceLevel.HIGH
            else -> FSQConfidenceLevel.NONE
        }
        return FSQVisit(
            id = visit.getVisitId(),
            location = convertFoursquareLocation(visit.location),
            locationType = locationType,
            confidence = confidence,
            arrivalTime = visit.arrival,
            venue = visit.getVenue()?.let { convertVenue(it) },
            otherPossibleVenues = visit.getOtherPossibleVenues().map { convertVenue(it) },
            hasDeparted = visit.hasDeparted(),
            departureTime = if (visit.hasDeparted()) visit.getDeparture() else null
        )
    }

    private fun convertVenue(venue: Venue): FSQVenue {
        return FSQVenue(
            id = venue.id,
            name = venue.name,
            locationInformation = convertVenueLocation(venue.location),
            partnerVenueId = venue.partnerVenueId,
            probability = venue.probability,
            chains = venue.venueChains.map { convertChain(it) },
            categories = venue.categories.map { convertCategory(it) },
            hierarchy = venue.hierarchy.map { convertVenueParent(it) }
        )
    }

    private fun convertVenueLocation(location: Venue.Location): FSQLocationInformation {
        return FSQLocationInformation(
            address = location.address,
            crossStreet = location.crossStreet,
            city = location.city,
            state = location.state,
            postalCode = location.postalCode,
            country = location.country,
            location = FSQLocation(
                latitude = location.lat,
                longitude = location.lng
            )
        )
    }

    private fun convertCategory(category: Category): FSQCategory {
        return FSQCategory(
            id = category.id ?: "",
            name = category.name ?: "",
            pluralName = category.pluralName,
            shortName = category.shortName,
            icon = category.image?.let { convertCategoryIcon(it) },
            isPrimary = category.isPrimary
        )
    }

    private fun convertCategoryIcon(photo: Photo): FSQCategoryIcon {
        return FSQCategoryIcon(
            prefix = photo.prefix ?: "",
            suffix = photo.suffix ?: ""
        )
    }

    private fun convertChain(chain: Venue.VenueChain): FSQChain {
        return FSQChain(
            id = chain.id ?: "",
            name = chain.name ?: ""
        )
    }

    private fun convertVenueParent(parent: Venue.VenueParent): FSQVenueParent {
        return FSQVenueParent(
            id = parent.id ?: "",
            name = parent.name ?: "",
            categories = parent.categories.map { convertCategory(it) }
        )
    }

    private fun convertGeofenceEvent(event: GeofenceEvent): FSQGeofenceEvent {
        val eventType = when (event.geofenceEventType) {
            GeofenceEventType.ENTRANCE -> FSQGeofenceEventType.ENTRANCE
            GeofenceEventType.DWELL -> FSQGeofenceEventType.DWELL
            GeofenceEventType.EXIT -> FSQGeofenceEventType.EXIT
            GeofenceEventType.PRESENCE -> FSQGeofenceEventType.PRESENCE
            else -> FSQGeofenceEventType.ENTRANCE
        }
        return FSQGeofenceEvent(
            geofenceId = event.id ?: "",
            name = event.name ?: "",
            eventType = eventType,
            venueId = event.venue?.id,
            venue = event.venue?.let { convertVenue(it) },
            partnerVenueId = event.partnerVenueId,
            location = FSQLocation(
                latitude = event.lat,
                longitude = event.lng
            ),
            timestamp = event.timestamp
        )
    }

    private fun convertFoursquareLocation(location: FoursquareLocation): FSQLocation {
        return FSQLocation(
            latitude = location.lat,
            longitude = location.lng
        )
    }
}

File: ios/Classes/FoursquareMovementSdkPlugin.swift
// Copyright 2026 Atmosphere Innovations, Inc. All rights reserved.
// Use of this source code is governed by an MIT-style license that can be
// found in the LICENSE file.

import Flutter
import UIKit
import CoreLocation
import MovementSdk

// Type aliases to disambiguate SDK types from Pigeon-generated types
typealias SDKCurrentLocation = MovementSdk.CurrentLocation
typealias SDKVisit = MovementSdk.Visit
typealias SDKVenue = MovementSdk.Venue
typealias SDKVenueLocation = MovementSdk.Venue.Location
typealias SDKCategory = MovementSdk.Category
typealias SDKCategoryIcon = MovementSdk.Category.Icon
typealias SDKChain = MovementSdk.Venue.Chain
typealias SDKGeofenceEvent = MovementSdk.GeofenceEvent

public class FoursquareMovementSdkPlugin: NSObject, FlutterPlugin, MovementSdkHostApi, MovementSdkManagerDelegate {
  private var flutterApi: MovementSdkFlutterApi?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let instance = FoursquareMovementSdkPlugin()
    MovementSdkHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: instance)
    instance.flutterApi = MovementSdkFlutterApi(binaryMessenger: registrar.messenger())
    MovementSdkManager.shared.delegate = instance
  }

  // MARK: - MovementSdkManagerDelegate

  public func movementManager(_ movementManager: MovementSdkManager, handle visit: MovementSdk.Visit) {
    let converted = convertVisit(visit)
    DispatchQueue.main.async { [weak self] in
      self?.flutterApi?.onVisit(visit: converted) { _ in }
    }
  }

  public func movementManager(_ movementManager: MovementSdkManager, handleBackfill visit: MovementSdk.Visit) {
    let converted = convertVisit(visit)
    DispatchQueue.main.async { [weak self] in
      self?.flutterApi?.onBackfillVisit(visit: converted) { _ in }
    }
  }

  public func movementManager(_ movementManager: MovementSdkManager, handle geofenceEvents: [MovementSdk.GeofenceEvent]) {
    let converted = geofenceEvents.map { convertGeofenceEvent($0) }
    DispatchQueue.main.async { [weak self] in
      self?.flutterApi?.onGeofenceEvents(events: converted) { _ in }
    }
  }

  // MARK: - MovementSdkHostApi Implementation

  func getInstallId(completion: @escaping (Result<String, any Error>) -> Void) {
    completion(.success(MovementSdkManager.shared.installId ?? ""))
  }

  func start() throws {
    MovementSdkManager.shared.start()
  }

  func stop() throws {
    MovementSdkManager.shared.stop()
  }

  func getCurrentLocation(completion: @escaping (Result<FSQCurrentLocation, any Error>) -> Void) {
    MovementSdkManager.shared.getCurrentLocation { sdkCurrentLocation, error in
      if let error = error {
        completion(.failure(PigeonError(code: "GET_CURRENT_LOCATION_ERROR", message: error.localizedDescription, details: nil)))
        return
      }
      guard let sdkCurrentLocation = sdkCurrentLocation else {
        completion(.failure(PigeonError(code: "GET_CURRENT_LOCATION_ERROR", message: "No current location available", details: nil)))
        return
      }
      completion(.success(self.convertCurrentLocation(sdkCurrentLocation)))
    }
  }

  func isEnabled(completion: @escaping (Result<Bool, any Error>) -> Void) {
    completion(.success(MovementSdkManager.shared.isEnabled))
  }

  func fireTestVisit(latitude: Double, longitude: Double) throws {
    let location = CLLocation(latitude: latitude, longitude: longitude)
    MovementSdkManager.shared.visitTester.fireTestVisit(location: location)
  }

  func showDebugScreen() throws {
    MovementSdkManager.shared.isDebugLogsEnabled = true
    guard #available(iOS 14.0, *) else { return }
    let viewController = UIApplication.shared.connectedScenes
      .compactMap { $0 as? UIWindowScene }
      .flatMap { $0.windows }
      .first { $0.isKeyWindow }?
      .rootViewController
    if let viewController = viewController {
      MovementSdkManager.shared.presentDebugViewController(parentViewController: viewController)
    }
  }

  func getUserInfo(completion: @escaping (Result<[String: String], any Error>) -> Void) {
    completion(.success(MovementSdkManager.shared.userInfo.source))
  }

  func setUserInfo(userInfo: [String: String], persisted: Bool) throws {
    let fsqUserInfo = UserInfo()
    for (key, value) in userInfo {
      fsqUserInfo.setUserInfo(value, forKey: key)
    }
    MovementSdkManager.shared.setUserInfo(fsqUserInfo, persisted: persisted)
  }

  func setUserId(userId: String, persisted: Bool) throws {
    let fsqUserInfo = MovementSdkManager.shared.userInfo
    fsqUserInfo.setUserId(userId)
    MovementSdkManager.shared.setUserInfo(fsqUserInfo, persisted: persisted)
  }

  func clearAllData() throws {
    MovementSdkManager.shared.clearAllData(completion: nil)
  }

  func getActiveVisit(completion: @escaping (Result<FSQVisit?, any Error>) -> Void) {
    MovementSdkManager.shared.getCurrentLocation { currentLocation, error in
      guard let currentLocation = currentLocation, error == nil else {
        completion(.success(nil))
        return
      }
      completion(.success(self.convertVisit(currentLocation.currentPlace)))
    }
  }

  // MARK: - Conversion Methods

  private func convertCurrentLocation(_ sdkCurrentLocation: SDKCurrentLocation) -> FSQCurrentLocation {
    return FSQCurrentLocation(
      currentPlace: convertVisit(sdkCurrentLocation.currentPlace),
      matchedGeofences: sdkCurrentLocation.matchedGeofences.map { convertGeofenceEvent($0) }
    )
  }

  private func convertVisit(_ sdkVisit: SDKVisit) -> FSQVisit {
    let locationType: FSQLocationType
    switch sdkVisit.locationType {
    case .home:
      locationType = .home
    case .work:
      locationType = .work
    case .venue:
      locationType = .venue
    default:
      locationType = .unknown
    }

    let confidence: FSQConfidenceLevel
    switch sdkVisit.confidence {
    case .low:
      confidence = .low
    case .medium:
      confidence = .medium
    case .high:
      confidence = .high
    default:
      confidence = .none
    }

    return FSQVisit(
      id: sdkVisit.id,
      location: sdkVisit.arrivalLocation.map { convertCLLocation($0) },
      locationType: locationType,
      confidence: confidence,
      arrivalTime: sdkVisit.arrivalDate.map { Int64($0.timeIntervalSince1970 * 1000) },
      venue: sdkVisit.venue.map { convertVenue($0) },
      otherPossibleVenues: sdkVisit.otherPossibleVenues.map { convertVenue($0) },
      hasDeparted: sdkVisit.hasDeparted,
      departureTime: sdkVisit.departureDate.map { Int64($0.timeIntervalSince1970 * 1000) }
    )
  }

  private func convertVenue(_ sdkVenue: SDKVenue) -> FSQVenue {
    return FSQVenue(
      id: sdkVenue.id,
      name: sdkVenue.name,
      locationInformation: sdkVenue.locationInformation.map { convertVenueLocation($0) },
      partnerVenueId: sdkVenue.partnerVenueId,
      probability: sdkVenue.probability?.doubleValue,
      chains: sdkVenue.chains.map { convertChain($0) },
      categories: sdkVenue.categories.map { convertCategory($0) },
      hierarchy: sdkVenue.hierarchy.map { convertVenueParent($0) }
    )
  }

  private func convertVenueLocation(_ sdkLocation: SDKVenueLocation) -> FSQLocationInformation {
    return FSQLocationInformation(
      address: sdkLocation.address,
      crossStreet: sdkLocation.crossStreet,
      city: sdkLocation.city,
      state: sdkLocation.state,
      postalCode: sdkLocation.postalCode,
      country: sdkLocation.country,
      location: FSQLocation(
        latitude: sdkLocation.coordinate.latitude,
        longitude: sdkLocation.coordinate.longitude
      )
    )
  }

  private func convertCategory(_ sdkCategory: SDKCategory) -> FSQCategory {
    return FSQCategory(
      id: sdkCategory.id ?? "",
      name: sdkCategory.name ?? "",
      pluralName: sdkCategory.pluralName,
      shortName: sdkCategory.shortName,
      icon: sdkCategory.icon.map { convertCategoryIcon($0) },
      isPrimary: sdkCategory.isPrimary
    )
  }

  private func convertCategoryIcon(_ sdkIcon: SDKCategoryIcon) -> FSQCategoryIcon {
    return FSQCategoryIcon(
      prefix: sdkIcon.prefix ?? "",
      suffix: sdkIcon.suffix ?? ""
    )
  }

  private func convertChain(_ sdkChain: SDKChain) -> FSQChain {
    return FSQChain(
      id: sdkChain.id ?? "",
      name: sdkChain.name ?? ""
    )
  }

  private func convertVenueParent(_ sdkParentVenue: SDKVenue) -> FSQVenueParent {
    return FSQVenueParent(
      id: sdkParentVenue.id,
      name: sdkParentVenue.name,
      categories: sdkParentVenue.categories.map { convertCategory($0) }
    )
  }

  private func convertGeofenceEvent(_ sdkEvent: SDKGeofenceEvent) -> FSQGeofenceEvent {
    let eventType: FSQGeofenceEventType
    switch sdkEvent.eventType {
    case .entrance:
      eventType = .entrance
    case .dwell:
      eventType = .dwell
    case .exit:
      eventType = .exit
    case .presence:
      eventType = .presence
    default:
      eventType = .entrance
    }

    return FSQGeofenceEvent(
      geofenceId: sdkEvent.geofenceId,
      name: sdkEvent.name,
      eventType: eventType,
      venueId: sdkEvent.venue?.id,
      venue: sdkEvent.venue.map { convertVenue($0) },
      partnerVenueId: sdkEvent.partnerVenueId,
      location: FSQLocation(
        latitude: sdkEvent.location.coordinate.latitude,
        longitude: sdkEvent.location.coordinate.longitude
      ),
      timestamp: Int64(sdkEvent.timestamp.timeIntervalSince1970 * 1000)
    )
  }

  private func convertCLLocation(_ location: CLLocation) -> FSQLocation {
    return FSQLocation(
      latitude: location.coordinate.latitude,
      longitude: location.coordinate.longitude
    )
  }
}
